File: sdk/src/main/java/com/meniga/sdk/models/challenges/operators/MenigaChallengesOperationsImp.java
package com.meniga.sdk.models.challenges.operators;

import com.meniga.sdk.MenigaSDK;
import com.meniga.sdk.helpers.MenigaDecimal;
import com.meniga.sdk.helpers.Result;
import com.meniga.sdk.models.challenges.MenigaChallenge;
import com.meniga.sdk.models.challenges.enums.CustomChallengeColor;
import com.meniga.sdk.webservices.requests.AcceptChallenge;
import com.meniga.sdk.webservices.requests.CreateChallenge;
import com.meniga.sdk.webservices.requests.CreateChallengeTypeData;
import com.meniga.sdk.webservices.requests.DeleteChallenge;
import com.meniga.sdk.webservices.requests.GetChallenge;
import com.meniga.sdk.webservices.requests.GetChallenges;
import com.meniga.sdk.webservices.requests.UpdateChallenge;

import org.joda.time.DateTime;

import java.util.List;
import java.util.UUID;

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
public class MenigaChallengesOperationsImp implements MenigaChallengesOperations {

	@Override
	public Result<List<MenigaChallenge>> getChallenges(boolean includeExpired,
	                                                   boolean excludeSuggested,
	                                                   boolean excludeAccepted) {
		GetChallenges req = new GetChallenges();
		req.includeExpired = includeExpired;
		req.excludeSuggested = excludeSuggested;
		req.excludeAccepted = excludeSuggested;
		return MenigaSDK.executor().getChallenges(req);
	}

	@Override
	public Result<MenigaChallenge> getChallenge(UUID id) {
		GetChallenge req = new GetChallenge();
		req.id = id;

		return MenigaSDK.executor().getChallenge(req);
	}

	@Override
	public Result<MenigaChallenge> acceptChallenge(UUID id, MenigaDecimal targetAmount) {
		AcceptChallenge req = new AcceptChallenge();
		req.id = id;
		req.targetAmount = targetAmount;
		return MenigaSDK.executor().acceptChallenge(req);
	}

	@Override
	public Result<MenigaChallenge> createChallenge(String title, String description,
	                                                   DateTime startDate, DateTime endDate,
	                                                   List<Long> categoryIds, MenigaDecimal targetAmount,
	                                                   Long iconId, CustomChallengeColor color) {
		CreateChallenge req = new CreateChallenge();
		req.title = title;
		req.description = description;
		req.startDate = startDate;
		req.endDate = endDate;
		req.typeData = new CreateChallengeTypeData();
		req.typeData.categoryIds = categoryIds;
		req.typeData.targetAmount = targetAmount;
		req.iconId = iconId;
		if (color != null) {
			req.typeData.metaData = "{\"color\": \"" + color.toColorString() + "\"}";
		}

		return MenigaSDK.executor().createChallenge(req);
	}

	@Override
	public Result<Void> updateChallenge(MenigaChallenge challenge, boolean updateCategoryIds) {
		UpdateChallenge req = new UpdateChallenge();
		req.id = challenge.getId();

		/*req.title = challenge.getTitle();
		req.description = challenge.getDescription();
		req.startDate = challenge.getStartDate();
		req.endDate = challenge.getEndDate();
		req.iconUrl = challenge.getIconUrl();*/

		req.typeData = new UpdateChallenge.TypeData();
		req.typeData.targetAmount = challenge.getTargetAmount();

		if (updateCategoryIds) {
			//req.typeData.categoryIds = challenge.getCategoryIds();
		}
		if (challenge.getCustomChallengeColor() != null) {
			//req.typeData.metaData = "{\"color\": \"" + challenge.getCustomChallengeColor().toColorString() + "\"}";
		}

		return MenigaSDK.executor().updateChallenge(req);
	}

	@Override
	public Result<Void> deleteChallenge(UUID id) {
		DeleteChallenge req = new DeleteChallenge();
		req.id = id;
		return MenigaSDK.executor().deleteChallenge(req);
	}
}

File: sdk/src/main/java/com/meniga/sdk/models/challenges/operators/MenigaChallengesOperationsImp.kt
/*
 * Copyright 2017-2018 Meniga Iceland Inc.
 */
package com.meniga.sdk.models.challenges.operators

import com.meniga.sdk.MenigaSDK
import com.meniga.sdk.helpers.MenigaDecimal
import com.meniga.sdk.helpers.Result
import com.meniga.sdk.models.challenges.CategoryDefinition
import com.meniga.sdk.models.challenges.FetchChallengeFilter
import com.meniga.sdk.models.challenges.MenigaChallenge
import com.meniga.sdk.models.challenges.NewChallenge
import com.meniga.sdk.webservices.challenge.AcceptChallenge
import com.meniga.sdk.webservices.challenge.CreateChallenge
import com.meniga.sdk.webservices.challenge.DeleteChallenge
import com.meniga.sdk.webservices.challenge.GetChallenge
import com.meniga.sdk.webservices.challenge.GetChallengeHistory
import com.meniga.sdk.webservices.challenge.GetChallenges
import com.meniga.sdk.webservices.challenge.UpdateChallenge
import java.util.UUID

class MenigaChallengesOperationsImp : MenigaChallengesOperations {

    override fun getChallenges(includeExpired: Boolean,
                               excludeSuggested: Boolean,
                               excludeAccepted: Boolean): Result<List<MenigaChallenge>> {
        val req = GetChallenges()
        req.includeExpired = includeExpired
        req.excludeSuggested = excludeSuggested
        req.excludeAccepted = excludeSuggested
        return MenigaSDK.executor().getChallenges(req)
    }

    override fun getChallenges(filter: FetchChallengeFilter): Result<List<MenigaChallenge>> {
        val request = GetChallenges(
                filter.includeExpired,
                filter.excludeSuggested,
                filter.excludeAccepted,
                filter.includeDisabled)
        return MenigaSDK.executor().getChallenges(request)
    }

    override fun getChallenge(id: UUID): Result<MenigaChallenge> {
        val req = GetChallenge(id)
        return MenigaSDK.executor().getChallenge(req)
    }

    override fun acceptChallenge(id: UUID, targetAmount: MenigaDecimal): Result<MenigaChallenge> {
        val req = AcceptChallenge()
        req.id = id
        req.targetAmount = targetAmount
        return MenigaSDK.executor().acceptChallenge(req)
    }

    override fun disableChallenge(id: UUID): Result<Void> {
        return MenigaSDK.executor().disableChallenge(id)
    }

    override fun enableChallenge(id: UUID): Result<Void> {
        return MenigaSDK.executor().enableChallenge(id)
    }

    override fun createChallenge(challenge: NewChallenge): Result<MenigaChallenge> {
        val typeData = CreateChallenge.CreateChallengeTypeData()
        challenge.categoryDefinition.let {
            when (it) {
                is CategoryDefinition.CategoryList -> typeData.categoryIds = it.categoryIds
                is CategoryDefinition.Type -> typeData.categoryType = it.categoryType
            }
        }
        typeData.targetAmount = challenge.targetAmount
        typeData.recurringInterval = challenge.recurringInterval
        if (challenge.customChallengeColor != null) {
            typeData.metaData = "{\"color\": \"" + challenge.customChallengeColor!!.toColorString() + "\"}"
        }

        val req = CreateChallenge(
                challenge.title,
                challenge.description,
                challenge.startDate,
                challenge.endDate,
                challenge.iconUrl,
                typeData
        )
        return MenigaSDK.executor().createChallenge(req)
    }

    override fun updateChallenge(challenge: MenigaChallenge, updateCategoryIds: Boolean): Result<Void> {
        val req = UpdateChallenge()
        req.id = challenge.id

        req.title = challenge.title
        req.description = challenge.description
        req.startDate = challenge.startDate
        req.endDate = challenge.endDate
        req.iconUrl = challenge.iconUrl

        req.typeData = UpdateChallenge.TypeData()
        req.typeData!!.targetAmount = challenge.targetAmount

        if (updateCategoryIds) {
            req.typeData!!.categoryIds = challenge.categoryIds
        }
        if (challenge.customChallengeColor != null) {
            req.typeData!!.metaData = "{\"color\": \"" + challenge.customChallengeColor.toColorString() + "\"}"
        }

        return MenigaSDK.executor().updateChallenge(req)
    }

    override fun deleteChallenge(id: UUID): Result<Void> {
        val req = DeleteChallenge(id)
        return MenigaSDK.executor().deleteChallenge(req)
    }

    override fun getChallengeHistory(id: UUID, page: Int, numPerPage: Int): Result<List<MenigaChallenge>> {
        val skip = if (page == 0) 0 else numPerPage / page
        val req = GetChallengeHistory(id, skip, numPerPage)
        return MenigaSDK.executor().getChallengeHistory(req)
    }
}
